src/lib/services/adapters/xmpp/plugins/multi-user-chat.plugin.ts

```diff
@@ -233,27 +265,109 @@ export class MultiUserChatPlugin extends AbstractXmppPlugin {
             return this.handleRoomPresenceStanza(stanza);
         } else if (this.isRoomMessageStanza(stanza)) {
             return this.handleRoomMessageStanza(stanza, archiveDelayElement);
+        } else if (this.isRoomSubjectStanza(stanza)) {
+            return this.handleRoomSubjectStanza(stanza, archiveDelayElement);
+        } else if (this.isRoomInviteStanza(stanza)) {
+            return this.handleRoomInviteStanza(stanza);
         }
         return false;
     }
 
     private isRoomPresenceStanza(stanza: Stanza): boolean {
         return stanza.name === 'presence' && (
-            stanza.getChild('x', 'http://jabber.org/protocol/muc')
-            || stanza.getChild('x', 'http://jabber.org/protocol/muc#user')
+            stanza.getChild('x', MultiUserChatPlugin.MUC)
+            || stanza.getChild('x', MultiUserChatPlugin.MUC_USER)
         ) != null;
     }
 
     private handleRoomPresenceStanza(stanza: Stanza): boolean {
         const handleStanza = this.roomJoinResponseHandlers.get(stanza.attrs.from);
+        const stanzaType = stanza.attrs.type;
+
+        if (stanzaType === 'error') {
+            this.handlerErrorMessage(stanza);
+        }
+
         if (handleStanza) {
             this.roomJoinResponseHandlers.delete(stanza.attrs.from);
             handleStanza(stanza);
-            return true;
+        }
+
+        const roomJid = parseJid(stanza.attrs.from.split('/')[0]);
+
+        const x = stanza.getChild('x', MultiUserChatPlugin.MUC_USER);
+
+        const occupant: Occupant = {
+            jid: x.getChild('item').attrs.jid,
+            affiliation: x.getChild('item').attrs.affiliation,
+            role: x.getChild('item').attrs.role,
+            nick: stanza.attrs.from.split('/')[1],
+        };
+
+        if (stanzaType === 'unavailable') {
+            const statusCodes = x.getChildren('status').map(status => status.attrs.code);
+
+            if (statusCodes.includes('307') && occupant.affiliation === Affiliation.none) {
+                const actor = x.getChild('item').getChild('actor')?.attrs.nick;
+                const reason = x.getChild('item').getChild('reason')?.getText();
+
+                return this.handleOccupantKicked(occupant, roomJid.toString(), statusCodes.includes('110'), actor, reason);
+            } else if (statusCodes.includes('301') && occupant.affiliation === Affiliation.outcast) {
+                return this.handleOccupantBanned(occupant);
+            } else if (statusCodes.includes('303')) {
+                return this.handleOccupantChangedNick(occupant, x.getChild('item').attrs.nick);
+            } else {
+                return this.handleOccupantLeft(occupant, roomJid.toString(), statusCodes.includes('110'));
+            }
+        } else if (!stanzaType) {
+            const room = this.getOrCreateRoom(roomJid);
+            return this.handleOccupantJoined(occupant, room.roomJid.toString());
```

Comment:
I think we should differentiate between a join and an affiliation/role change when propagating occupant changes.

Reply:
The problem here is that I can't differentiate between join and occupant changes. The stanzas given are almost the same.

Join stanza:
```xml
<presence
        from='coven@chat.shakespeare.lit/thirdwitch'
        id='27C55F89-1C6A-459A-9EB5-77690145D624'
        to='crone1@shakespeare.lit/desktop'>
    <x xmlns='http://jabber.org/protocol/muc#user'>
        <item affiliation='member'
              role='participant'/>
    </x>
</presence>
```
Granting membership:
```xml
<presence
        from='coven@chat.shakespeare.lit/thirdwitch'
        to='crone1@shakespeare.lit/desktop'>
    <x xmlns='http://jabber.org/protocol/muc#user'>
        <item affiliation='member'
              jid='hag66@shakespeare.lit/pda'
              role='participant'
              nick='thirdwitch'/>
    </x>
</presence>
```

The id on the join stanza isn't given like in the example above. So should I check if the occupant is already in the room to differentiate between these stanzas?

Reply:
Yes, I think this is the way to go. If you already know this occupant, then emit a 'modified' change, possibly by diffing the fields and finding out what exactly changed. Otherwise it's an 'joined' change.